### PR TITLE
[inductor] make inductor work with new triton compile interface

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115881
* __->__ #115878

Recent 2 triton PRs (https://github.com/openai/triton/pull/2701, https://github.com/openai/triton/pull/2756) change the interface for triton.compile, this PR added the necessary change on inductor side to work with both old and new compile API.

Also there is some simplification between compilation call in subprocess and the one in main process
- previously we pass warm_cache_only=True if the compilation happens in subprocess. But triton never use that argument in the currently used pin. So I removed that
- previously we only pass compute_capability if compilation happens in subprocess. The PR change that to always passing compute_capability to triton.compile no matter if the compilation happens in main or sub process.

Updated:
There are more interface change from triton side. E.g.
- tl.math.{min, max} now requires a propagate_nan argument
- JITFunction.run now requires a warmup argument. This affect the benchmarking phase of matmul max-autotune; on the other hand, JITFunction.run forbids stream argument now. Simply removing passing this in when benchmarking matmul triton kernel will work for both old and new version of triton.
- triton Autotuner change attribute name from 'warmup' to 'num_warmup' and from 'rep' to 'num_rep'. This cause dynamo failed to handle triton Autotuner object since dynamo TritonKernelVariable makes assumption about attribute names. It's used in some test cases that a model call triton Autotuner directly.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
Looks like there are some other changes needed, see `TypeError("max() missing 1 required positional argument: 'propagate_nan'")` in the next pr

File: torch/_inductor/triton_heuristics.py [hunk 1]
```diff
@@ -329,14 +363,29 @@ def launcher({', '.join(def_args)}, grid, stream):
                 else:
                     grid_0, grid_1, grid_2 = grid
 
+                if hasattr(bin, "run"):
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/triton_heuristics.py:366]: hmm, this fail if bin does not have 'c_wrapper' attribute? The attribute is removed in newer triton.

```
>>> class Cls: pass
...
>>> o = Cls()
>>> o.a = 3
>>> getattr(o, "a", o.b) # this fail. Look like o.b is evaluated even if o has attribute 'a' and o.b is not used
```

[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/triton_heuristics.py:366]: oh, true.. nvm. `getattr(bin, "run", getattr(bin, "c_wrapper", None))`. Probably not worth it though.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -2099,6 +2099,10 @@ def codegen_kernel(self, name=None):
                     from torch._inductor.ir import TileHint
                     from torch._inductor.triton_heuristics import AutotuneHint, {heuristics}
                     from torch._inductor.utils import instance_descriptor
+                    try:
+                        from triton.compiler.compiler import AttrsDescriptor
+                    except ImportError:
+                        pass  # ignore
```
Comment (jansel) [inline torch/_inductor/codegen/triton.py:2105]: Let's check the Triton version either include or exclude this to keep code cleaner.

File: torch/_inductor/triton_heuristics.py [hunk 3]
```diff
@@ -329,14 +363,29 @@ def launcher({', '.join(def_args)}, grid, stream):
...
-                                stream, bin.cu_function, None, None, None,
+                    runner(grid_0, grid_1, grid_2, bin.num_warps,
+                                bin.num_ctas, *cluster_dims, bin.shared,
+                                stream, function, None, None, None,
                                 {', '.join(call_args)})
                 else:
-                    bin.c_wrapper(grid_0, grid_1, grid_2, bin.num_warps, bin.shared,
-                                stream, bin.cu_function, None, None, None,
+                    runner(grid_0, grid_1, grid_2, bin.num_warps, bin.shared,
+                                stream, function, None, None, None,
```
Comment (jansel) [inline torch/_inductor/triton_heuristics.py:388]: Let's do the version check at compile time and emit smaller/faster launcher code.  No need to call `hasattr` a bunch of times on every single kernel launch.

**Review by jansel — CHANGES_REQUESTED** [review]

**Review by jansel — APPROVED** [review]

**Comment by kit1980** [comment]
This broke ROCm tests:
> torch._dynamo.exc.BackendCompilerFailed: backend='compile_fx_wrapper' raised:
> NotImplementedError: No interface for device hip

https://github.com/pytorch/pytorch/actions/runs/7282149837/job/19844618618

**Comment by kit1980** [comment]
@shunting314 Fix the typo in the PR title while you're at it?

**Pull request closed without merging**